### PR TITLE
Improvement: Synchronise zoom action 

### PR DESCRIPTION
**Pull request by meiliang1123**
- Synchronise zoom action between editor and process tab of laser / cnc
- Don't show cnc-warning popup on next open when warning label checked

### CODE REVIEW

File: src/app/components/SMCanvas/Canvas.jsx [hunk 1]
```diff
@@ -101,6 +108,21 @@ class Canvas extends Component {
         window.addEventListener('resize', this.resizeWindow, false);
     }
 
+    // just for laser and cnc, dont set scale prop for 3dp
+    componentWillReceiveProps(nextProps) {
+        if (nextProps.scale !== this.lastScale) {
+            const currentScale = this.initialDistance / (this.camera.position.distanceTo(this.controls.target));
+            this.controls.setScale(currentScale / nextProps.scale);
+            this.lastScale = nextProps.scale;
```
Comment (parachvte) [inline src/app/components/SMCanvas/Canvas.jsx:116]: First lastScale update

File: src/app/components/SMCanvas/Canvas.jsx [hunk 2]
```diff
@@ -109,6 +131,23 @@ class Canvas extends Component {
         this.transformControls && this.transformControls.dispose();
     }
 
+    onScale = () => {
+        const currentScale = this.initialDistance / (this.camera.position.distanceTo(this.controls.target));
+        if (Math.abs(currentScale - this.lastScale) > EPS) {
+            this.lastScale = currentScale;
```
Comment (parachvte) [inline src/app/components/SMCanvas/Canvas.jsx:137]: Second lastScale update

We only need one?

File: src/app/components/SMCanvas/Canvas.jsx [hunk 3]
```diff
@@ -109,6 +131,23 @@ class Canvas extends Component {
...
+            this.props.updateScale(currentScale);
+        }
+    }
```
Comment (parachvte) [inline src/app/components/SMCanvas/Canvas.jsx:140]: ;

File: src/app/components/SMCanvas/Canvas.jsx [hunk 4]
```diff
@@ -109,6 +131,23 @@ class Canvas extends Component {
...
+
+    onChangeTarget = () => {
+        clearTimeout(this.timeoutChangeTarget);
+        this.timeoutChangeTarget = setTimeout(() => {
+            const { x, y } = this.controls.target;
+            this.lastTarget = { x, y };
+            this.props.updateTarget(this.lastTarget);
+        }, 200);
+    }
```
Comment (parachvte) [inline src/app/components/SMCanvas/Canvas.jsx:149]: ;

File: src/app/components/SMCanvas/Canvas.jsx [hunk 5]
```diff
@@ -109,6 +131,23 @@ class Canvas extends Component {
...
+        const currentScale = this.initialDistance / (this.camera.position.distanceTo(this.controls.target));
+        if (Math.abs(currentScale - this.lastScale) > EPS) {
+            this.lastScale = currentScale;
+            this.props.updateScale(currentScale);
+        }
+    }
+
+    onChangeTarget = () => {
+        clearTimeout(this.timeoutChangeTarget);
+        this.timeoutChangeTarget = setTimeout(() => {
```
Comment (parachvte) [inline src/app/components/SMCanvas/Canvas.jsx:144]: Not elegant solution :(

File: src/app/components/SMCanvas/Controls.jsx [hunk 6]
```diff
@@ -445,17 +451,24 @@ class Controls extends EventEmitter {
...
+            this.target.add(this.panOffset);
+            this.panOffset.set(0, 0, 0);
+            this.onPan();
+        }
 
         // re-position camera
         this.camera.position.copy(this.target).add(this.offset);
         this.camera.lookAt(this.target);
 
+        if (this.scale !== 1) {
```
Comment (parachvte) [inline src/app/components/SMCanvas/Controls.jsx:467]: put it in L448?

**Review by parachvte — CHANGES_REQUESTED** [review]
This is not a good solution for zoom synchronization.

It works for now, and we will refactor it afterwards.

File: src/app/components/SMCanvas/Canvas.jsx (hunk 5 above)
Comment (meiliang1123) [reply src/app/components/SMCanvas/Canvas.jsx]: changed to trigger on mouse up

File: src/app/components/SMCanvas/Canvas.jsx (hunk 2 above)
Comment (meiliang1123) [reply src/app/components/SMCanvas/Canvas.jsx:137]: Set lastScale here to avoid update loop, app will crash if this line removed

**Pull request merged**